docs/Chinese/config.md
---
sort: 2
---
# 配置项描述

<a id="baseconfig"></a>
## 基本配置项
不管部署到什么平台，都先要正确配置config.py，这一节描述几个简单的配置项，其他配置项在下面章节详细描述。    

| 配置项              |    含义                                                  |
| ------------------- | -------------------------------------------------------- |
| APP_ID              | 应用标识符，gae平台为应用ID，其他平台用于标识数据库等资源    |
| APP_DOMAIN          | 应用部署后的域名                                          |
| KE_TEMP_DIR         | 制作电子书时的临时目录，为空则使用内存保存临时文件           |
| EBOOK_SAVE_DIR      | 如果需要在线阅读功能，配置此目录用于保存电子书              |
| DOWNLOAD_THREAD_NUM | 下载网页的线程数量，需要目标平台支持多线程，最大值为5        |
| ALLOW_SIGNUP        | 是否允许用户注册，"yes"-用户可以自主注册（可以通过邀请码限制），"no"- 只有管理员能创建其他账号。|
| SECRET_KEY          | 浏览器session的加密密钥，建议修改，任意字符串即可           |
| ADMIN_NAME          | 管理员的账号名                                            |
| POCKET_CONSUMER_KEY | 用于稍后阅读服务Pocket，可以用你自己的Key或就直接使用这个    |
| HIDE_MAIL_TO_LOCAL  | 是否允许将生成的邮件保存到本地，用于调试或测试目的           |




<a id="database"></a>
## 数据库选择
数据库用于保存应用的配置数据和订阅数据。    
得益于SQL数据库ORM库 [peewee](https://pypi.org/project/peewee/) 和作者因KindleEar需要而创建的NoSQL数据库ODM库 [weedata](https://github.com/cdhigh/weedata)，KindleEar支持很多数据库类型，包括：datastore, sqlite, mysql, postgresql, cockroachdb, mongodb, redis, pickle， 基本兼容了市面上的主流数据库，更适合全平台部署，平台支持什么数据库就可以使用什么数据库。    
如果目标平台同时支持SQL/NoSQL，则建议NoSQL，其最大的优点是万一以后升级需要修改数据库结构，则NoSQL不会影响原有数据，而SQL会删掉原数据。    
本应用的数据量不大，确切的来说，很小很小，一般就几十行数据，选择什么数据库都不会对资源消耗和性能造成什么影响，即使最简单的就用一个文本文件当做数据库使用可能都会比其他正规的数据库要快。    



### Datastore
Datastore为google的NoSQL数据库，我们要使用的是firebase的datastore模式。    
如果要部署到google cloud，基本上你只能选择datastore，因为它有免费额度。    
要使用datastore，参数配置如下：
```python
DATABASE_URL = 'datastore'
```



### SQLite
单文件数据库。适用于有本地文件系统读写权限的平台，特别是资源受限系统比如树莓派和各种派之类的。    
数据库文件的路径支持绝对路径和相对路径，相对路径的基目录为项目目录。  
要使用datastore，参数配置如下：
```python
#template:
DATABASE_URL = 'sqlite:////path/to/database.db'
#examples:
DATABASE_URL = 'sqlite:////C:/Users/name/kindleear/site.db'
DATABASE_URL = 'sqlite:////home/username/dbfilename.db'
DATABASE_URL = 'sqlite:///dbfilename.db'  #relative path
```



### MySQL/PostgreSQL/CockroachDB
典型企业级SQL数据库。大炮打蚊子，如果平台支持，直接使用也无妨。    
参数配置如下：
```python
#template:
DATABASE_URL = 'mysql://username:password@hostname:port/database_name'
DATABASE_URL = 'postgresql://username:password@hostname:port/database_name'

#examples:
DATABASE_URL = 'mysql://root:password@localhost:3306/mydatabase'
DATABASE_URL = 'mysql://user:pass123@example.com:3306/mydatabase'
DATABASE_URL = 'postgresql://postgres:password@localhost:5432/mydatabase'
DATABASE_URL = 'postgresql://user:pass123@example.com:5432/mydatabase'

import os
db_username = os.getenv('DB_USERNAME')
db_password = os.getenv('DB_PASSWORD')
db_host = os.getenv('DB_HOST')
db_port = os.getenv('DB_PORT')
db_name = os.getenv('DB_NAME')
database_url = f"mysql://{db_username}:{db_password}@{db_host}:{db_port}/{db_name}"
```



### MongoDB
应用最广的典型NoSQL数据库。      
1. 参数配置如下：
```python
#template:
DATABASE_URL = 'mongodb://username:password@hostname:port/'
#examples:
DATABASE_URL = 'mongodb://127.0.0.1:27017/'
DATABASE_URL = 'mongodb://user:pass123@example.com:27017/'
```

2. 如果部署的目标平台没有安装mongodb-server，可以参考 [官方文档](https://www.mongodb.com/docs/manual/installation/) 进行安装。   
这里是在ubuntu的安装方法：
```bash
sudo apt install gnupg curl
curl -fsSL https://www.mongodb.org/static/pgp/server-7.0.asc | \
   sudo gpg -o /usr/share/keyrings/mongodb-server-7.0.gpg \
   --dearmor
echo "deb [ arch=amd64,arm64 signed-by=/usr/share/keyrings/mongodb-server-7.0.gpg ] https://repo.mongodb.org/apt/ubuntu jammy/mongodb-org/7.0 multiverse" | sudo tee /etc/apt/sources.list.d/mongodb-org-7.0.list
sudo apt update
sudo apt install -y mongodb-org
sudo systemctl start mongod
sudo systemctl enable mongod
```


### Redis
可以持久化到磁盘的NoSQL内存数据库。如果目标系统已经安装并使用了redis用于任务队列，则直接使用redis可以省去安装其他数据库的资源消耗，但使用前要做好相关的redis持久化配置，避免丢失数据。     
1. 参数配置如下(db_number可以省略，如果是0，建议省略)：    
```python
DATABASE_URL = 'redis://[:password]@hostname:port/db_number'
DATABASE_URL = 'redis://127.0.0.1:6379/'
DATABASE_URL = 'redis://:password123@example.com:6379/1'
```

2. 如果不是的目标平台没有安装redis，可以参考 [官方文档]() 进行安装。  
这里是在ubuntu的安装方法：   
```bash
sudo apt update
sudo apt install redis-server
sudo systemctl start redis-server
sudo systemctl enable redis-server
```


### Pickle
作者使用Python的pickle数据持久化标准库创建的一个非常简单的单文件NoSQL"数据库"，可以用于资源特别受限的系统或用于测试目的。   
数据库文件的路径支持绝对路径和相对路径，相对路径的基目录为项目目录。  
参数配置如下：  
```python
#template:
DATABASE_URL = 'pickle:////path/to/database.db'
#examples:
DATABASE_URL = 'pickle:////C:/Users/name/kindleear/site.db'
DATABASE_URL = 'pickle:////home/username/dbfilename.db'
DATABASE_URL = 'pickle:///dbfilename.db'  #relative path
```



<a id="taskqueue"></a>
## 任务队列和定时任务选择
任务队列用于异步执行抓取网页内容、制作电子书、发送邮件等耗时任务。     
定时任务用于定时检查是否需要推送、清零过期推送记录等事项。     

### gae
如果要部署到google cloud，你只能选择gae。
```python
TASK_QUEUE_SERVICE = "gae"
TASK_QUEUE_BROKER_URL = ""
```



### apscheduler
比较轻量，最简配置可以不依赖redis和其他数据库，直接使用内存保存任务状态，只是有一定的丢失任务风险，在任务队列执行过程中掉电重新上电后原任务不会重新运行，只能等新的任务时间到达。     
如果要使用数据库持久化，支持sqlite/mysql/postgresql/mongodb/redis，可以配置为DATABASE_URL相同的值。
```python
TASK_QUEUE_SERVICE = "apscheduler"

TASK_QUEUE_BROKER_URL = "memory" #use memory store
#or
TASK_QUEUE_BROKER_URL = "redis://127.0.0.1:6379/"
#or
TASK_QUEUE_BROKER_URL = "sqlite:////home/username/dbfilename.db"
```

注意事项：
1. apscheduler 3.x不支持多进程（不管使用什么jobstore），配合 gunicorn 使用时需要使用多线程，比如设置 workers=1, threads=3，并且 preload_app=False。   




### celery
最著名的任务队列，支持多种后端技术，包括redis、mongodb、sql、共享目录等，如果要使用数据库保存任务状态，可以配置为DATABASE_URL相同的值。
```python
TASK_QUEUE_SERVICE = "celery"

TASK_QUEUE_BROKER_URL = "redis://127.0.0.1:6379/"
#or
TASK_QUEUE_BROKER_URL = "sqlite:////home/username/dbfilename.db"
#or
TASK_QUEUE_BROKER_URL = "file:///var/celery/results/" #results is a directory
TASK_QUEUE_BROKER_URL = "file:////?/C:/Users/name/results/" #keep the prefix 'file:////?/' if in windows
```



### rq
比celery稍轻量的任务队列，依赖redis，需要额外安装redis服务。    
```python
TASK_QUEUE_SERVICE = "celery"
TASK_QUEUE_BROKER_URL = "redis://127.0.0.1:6379/"
```



<a id="sendmail"></a>
## 邮件发送服务选择
为了更方便使用和规避一些免费额度的限制，邮件发送服务可以等部署完成后在网页上配置。    
* **gae**:
部署到google cloud时建议使用，额度足够，邮件大小也慷慨，单邮件最大31.5MB。     

* **sendgrid**:
部署到其他平台时建议使用，需要额外 [注册账号](https://sendgrid.com/) 并申请一个ApiKey，单邮件最大30MB。 
注：我一直无法正常注册sendgrid，不管换多少种方法多少个邮件地址都无法正常登录sendgrid，所以此功能我没有亲自测试，测试通过的朋友可以告知我。    

* **mailjet**:
[Mailjet](https://www.mailjet.com/)是另一个选项， 需要注册账号并且申请ApiKey和SecretKey，单邮件最大15MB。记得发送前确认自己的发件地址有没有在 [Sender addresses](https://app.mailjet.com/account/sender) 里面。     
测试过程中还发现一个问题，如果你的发件人邮件地址不是你注册mailjet的地址，则mailjet不会报错，只是对方永远收不到。所以如果使用mailjet发送邮件失败，请确认发件地址是否正确。    

* **SMTP**:
这个选项就灵活了，大部分的电子邮件服务平台都支持SMTP，只是很多平台对SMTP登录都有诸多限制，使用前请注意阅读相关说明，特别是大多数平台的SMTP密码都和正常的账号密码不一致。   
除了使用市场上已有的服务外，ubuntu等平台也很方便的使用postfix等部署一个自己的SMTP服务。   




<a id="wsgi"></a>
## WSGI协议容器
KindleEar使用Flask框架实现Web界面管理，入口点在 main.app （main.py文件内的app实例对象），
可以使用任何支持wsgi标准接口的Web服务器软件启动此app即可。    
要求不高的，使用Flask的调试服务器直接启动都可以。    
如果部署到google cloud，默认使用 Gunicorn，也可以任意切换为 uWSGI, Tornado, mod_wsgi 等。    
其他目标平台也是一样的做法，选择已有的或自己喜欢的就行。  



<a id="pip"></a>
## requirements.txt
KindleEar使用requirements.txt管理各种库的依赖，在各种平台部署时都可以一行命令就完成应用的环境配置
```bash
pip install -r requirments.txt
```
因各种配置组合较多，要手动配置requirements.txt比较复杂，容易出错。    
为此，作者提供了一个脚本文件tools/update_req.py，
配置好config.py后，直接执行此文件就可以生成requirements.txt。    
或者你可以不用此脚本，而是将requirements.txt里面的所有注释都删除，也就是安装全部依赖库，反正也占用不了多少空间。




docs/English/config.md
---
sort: 2
---
# Configuration

<a id="baseconfig"></a>
## Base config
Regardless of the platform you deploy to, the first step is to correctly configure config.py.     
This section describes several simple configuration items, with more detailed descriptions of other configuration items in the subsequent sections.     

| Configuration Item | Meaning                                                |
| ------------------ | ------------------------------------------------------ |
| APP_ID             | Application identifier; for GAE platform, it's the app ID, while for other platforms, it's used to identify database and other resources |
| APP_DOMAIN         | Domain name of the deployed application                 |
| KE_TEMP_DIR        | Temporary directory for creating eBooks; if empty, temporary files are stored in memory |
| EBOOK_SAVE_DIR     | If online reading is required, this directory is used to store the generated e-books |
| DOWNLOAD_THREAD_NUM| Number of threads for downloading web pages; the target platform needs to support multithreading, with a maximum value of 5 |
| ALLOW_SIGNUP       | Whether to allow user registration; "yes" - Users can register autonomously (with the option to restrict via invitation codes), "no" - means account creation is done by administrators |
| SECRET_KEY         | Encryption key for browser session, recommended to change, any string is acceptable |
| ADMIN_NAME         | Administrator's username                                |
| POCKET_CONSUMER_KEY| Used for Pocket's read-later service; you can use your own key or use this one directly |
| HIDE_MAIL_TO_LOCAL| Whether to allow saving generated emails locally for debugging or testing purposes |




## Database Selection
The database is used to store application configuration data and subscription data.      
Thanks to the SQL database ORM library [peewee](https://pypi.org/project/peewee/) and the NoSQL database ODM library [weedata](https://github.com/cdhigh/weedata) created by the author for KindleEar, KindleEar supports many types of databases, including: datastore, sqlite, mysql, postgresql, cockroachdb, mongodb, redis, pickle.       
It basically covers the mainstream databases on the market and is more suitable for cross-platform deployment. You can use whatever database the platform supports.        
If the target platform supports both SQL and NoSQL, it is advisable to use NoSQL. Its major advantage is that in case of future upgrades requiring modifications to the database structure, NoSQL will not affect the existing data, whereas SQL would delete the original data.      
The amount of data in this application is very small, to be precise, very, very small, usually just a few dozen lines of data. Choosing any database will not have any impact on resource consumption and performance. Even using a simple text file as a database may be faster than other formal databases.      




### datastore
Datastore is Google's NoSQL database, and we will be using the Datastore mode of Firebase. If you want to deploy to Google Cloud, basically, you can only choose Datastore because it has free quotas.       
To use Datastore, the parameter configuration is as follows:    
```python
DATABASE_URL = 'datastore'
```



### SQLite
SQLite is a single-file database. It is suitable for platforms with local file system read and write permissions, especially resource-constrained systems such as Raspberry Pi and various derivatives.   
The database file path support both absolute path and relative path, with the project directory as the base directory for relative path.   
To use SQLite, the parameter configuration is as follows:    
```python
#template:
DATABASE_URL = 'sqlite:////path/to/database.db'
#examples:
DATABASE_URL = 'sqlite:////C:/Users/name/kindleear/site.db'
DATABASE_URL = 'sqlite:////home/username/dbfilename.db'
DATABASE_URL = 'sqlite:///dbfilename.db'  #relative path
```



### MySQL/PostgreSQL/CockroachDB
These are typical enterprise-level SQL databases. It's like using a cannon to kill a mosquito, but if the platform supports it, there's no harm in using them directly.      
Parameter configuration is as follows:   
```python
#template:
DATABASE_URL = 'mysql://username:password@hostname:port/database_name'
DATABASE_URL = 'postgresql://username:password@hostname:port/database_name'

#examples:
DATABASE_URL = 'mysql://root:password@localhost:3306/mydatabase'
DATABASE_URL = 'mysql://user:pass123@example.com:3306/mydatabase'
DATABASE_URL = 'postgresql://postgres:password@localhost:5432/mydatabase'
DATABASE_URL = 'postgresql://user:pass123@example.com:5432/mydatabase'

import os
db_username = os.getenv('DB_USERNAME')
db_password = os.getenv('DB_PASSWORD')
db_host = os.getenv('DB_HOST')
db_port = os.getenv('DB_PORT')
db_name = os.getenv('DB_NAME')
database_url = f"mysql://{db_username}:{db_password}@{db_host}:{db_port}/{db_name}"
```



### MongoDB
The most widely used typical NoSQL database.      
1. Parameter configuration is as follows:       
```python
#template:
DATABASE_URL = 'mongodb://username:password@hostname:port/'
#examples:
DATABASE_URL = 'mongodb://127.0.0.1:27017/'
DATABASE_URL = 'mongodb://user:pass123@example.com:27017/'
```

2. If mongodb is not installed on the target platform, you can refer to the [official documentation](https://www.mongodb.com/docs/manual/installation/) for installation instructions. Here is the installation method for Ubuntu:   
```bash
sudo apt install gnupg curl
curl -fsSL https://www.mongodb.org/static/pgp/server-7.0.asc | \
   sudo gpg -o /usr/share/keyrings/mongodb-server-7.0.gpg \
   --dearmor
echo "deb [ arch=amd64,arm64 signed-by=/usr/share/keyrings/mongodb-server-7.0.gpg ] https://repo.mongodb.org/apt/ubuntu jammy/mongodb-org/7.0 multiverse" | sudo tee /etc/apt/sources.list.d/mongodb-org-7.0.list
sudo apt update
sudo apt install -y mongodb-org
sudo systemctl start mongod
sudo systemctl enable mongod
```



### Redis
A memory NoSQL database that can persist to disk.    
If the target system already has Redis installed and used for task queues, using Redis directly can save the resource consumption of installing other databases.    
However, before using it, relevant Redis persistence configurations should be done to avoid data loss.    
1. Parameter configuration is as follows (the db_number can be omitted, but if it's 0, it's recommended to omit it):   
```python
DATABASE_URL = 'redis://[:password]@hostname:port/db_number'
DATABASE_URL = 'redis://127.0.0.1:6379/0'
DATABASE_URL = 'redis://:password123@example.com:6379/1'
```

2. If Redis is not installed on the target platform, you can refer to the [official documentation](https://redis.io/docs/install/install-redis/) for installation instructions. Here is the installation method for Ubuntu:   
```bash
sudo apt update
sudo apt install redis-server
sudo systemctl start redis-server
sudo systemctl enable redis-server
```



### Pickle
A very simple single-file NoSQL "database" created by the author using Python's pickle data persistence standard library.    
It can be used for resource-constrained systems or for testing purposes.    
The database file path support both absolute path and relative path, with the project directory as the base directory for relative path.   
Parameter configuration is as follows:   
```python
#template:
DATABASE_URL = 'pickle:////path/to/database.db'
#examples:
DATABASE_URL = 'pickle:////C:/Users/name/kindleear/site.db'
DATABASE_URL = 'pickle:////home/username/dbfilename.db'
DATABASE_URL = 'pickle:///dbfilename.db'  #relative path
```




<a id="taskqueue"></a>
## Task Queue and Scheduler Selection
Task queues are used for asynchronously executing tasks such as fetching web content, creating eBooks, sending emails, etc.    
Scheduler tasks are used for periodically checking whether there is a need for pushing notifications, resetting    expired push records, etc.    



### gae
If you want to deploy to Google Cloud, you can only choose GAE.    
```python
TASK_QUEUE_SERVICE = "gae"
TASK_QUEUE_BROKER_URL = ""
```



### apscheduler
Comparatively lightweight, with the simplest configuration, it can work without relying on Redis or other databases by directly using memory to store task states.   
However, there is a risk of losing tasks. If power is lost during the execution of a task, the original task will not rerun after power is restored; it will only wait for the next scheduled time.    
If database persistence is required, it supports SQLite/MySQL/PostgreSQL/MongoDB/Redis, and you can configure it with the same value as DATABASE_URL.    
```python
TASK_QUEUE_SERVICE = "apscheduler"

TASK_QUEUE_BROKER_URL = "memory" # use memory store
# or
TASK_QUEUE_BROKER_URL = "redis://127.0.0.1:6379/"
# or
TASK_QUEUE_BROKER_URL = "sqlite:////home/username/dbfilename.db"
```

Notes:
1. The apscheduler 3.x does not support multi-processes (No matter what job store is used). When used with gunicorn, using multi-threads instead, for example, setting workers=1, threads=3, and disabling preload_app.   



### celery
The most famous task queue, supports various backends such as Redis, MongoDB, SQL, shared directories, etc.     
If database persistence for task states is required, it can be configured with the same value as DATABASE_URL.   
```python
TASK_QUEUE_SERVICE = "celery"

TASK_QUEUE_BROKER_URL = "redis://127.0.0.1:6379/"
# or
TASK_QUEUE_BROKER_URL = "sqlite:////home/username/dbfilename.db"
# or
TASK_QUEUE_BROKER_URL = "file:///var/celery/results/" # results is a directory
TASK_QUEUE_BROKER_URL = "file:////?/C:/Users/name/results/" # keep the prefix 'file:////?/' if in windows
```



### rq
Slightly lighter-weight than Celery, it depends on Redis and requires an additional installation of the Redis service.  
```python
TASK_QUEUE_SERVICE = "celery"
TASK_QUEUE_BROKER_URL = "redis://127.0.0.1:6379/"
```




<a id="sendmail"></a>
## Email Sending Service Selection
To make it more convenient to use and avoid some limitations of free quotas, the email sending service can be configured on the web page after deployment is complete.    
* **GAE**:
Recommended for deployment to Google Cloud, with sufficient quotas and generous email size limit, with a maximum of 31.5MB per email.    

* **SendGrid**:
Recommended for deployment to other platforms, requires additional [registration](https://sendgrid.com/) and application for an API key, with a maximum of 30MB per email.    
Note: I have been unable to register with SendGrid successfully. Regardless of the methods used or the number of email addresses tried, I cannot login to SendGrid. Therefore, I have not personally tested this functionality. If any friends have successfully tested it, please inform me.

* **mailjet**:
You can also use [Mailjet](https://www.mailjet.com/). Just sign up for an account and get your ApiKey and SecretKey. The largest email you can send is 15MB. Don't forget to check that your sender address is on the [Sender addresses](https://app.mailjet.com/account/sender) list before you send any mail.    
During testing, another issue was discovered: if your sender email address is not the one you registered with Mailjet, Mailjet will not report an error; instead, the recipient will never receive the email. Therefore, if sending emails fails with Mailjet, please verify that the sender address is correct.   

* **SMTP**:
This option is flexible, as most email service platforms support SMTP. However, many platforms have various restrictions on SMTP. Before use, please carefully read the relevant instructions, especially considering that the SMTP password for most platforms differs from the regular account password.   

In addition to using existing services available in the market, platforms like Ubuntu also offer convenient ways to deploy your own SMTP service by using postfix.   





<a id="wsgi"></a>
## WSGI container
KindleEar uses the Flask framework to implement web interface management, with the entry point being `main.app` (the app instance object inside the main.py file).      
You can start this app using any web server software that supports the WSGI standard interface.     
For low requirements, you can directly start with the Flask debug server.     
If deploying to Google Cloud, it defaults to using Gunicorn, but you can freely switch to uWSGI, Tornado, mod_wsgi, etc.    
The same applies to other target platforms; choose whichever you prefer or are familiar with.    



<a id="pip"></a>
## requirements.txt
KindleEar uses requirements.txt to manage dependencies on various libraries.     
It allows for easy environment configuration with just one command on various platforms.    
```bash
pip install -r requirements.txt
```

Due to the various configuration combinations, manually configuring requirements.txt can be complex and prone to errors.   
To address this, the author provides a script file `tools/update_req.py`.    
After configuring config.py, simply execute this file to generate requirements.txt.   
Alternatively, you can choose not to use this script and remove all comments in requirements.txt, which installs all dependencies, as it won't take up much space anyway.   





changelog.md
# Changelog for KindleEar

## 1.26.8
  1. "网友分享" 添加一个搜索栏。（小彩蛋：输入 "#download" 可以下载全部RSS资源）
  2. 更新README和FAQ关于部署的说明。
  
## 1.26.7
  1. 适配 kindle 固件 5.9.x (在书籍下方显示时间而不是作者)。
  2. gufengmh地址变更。
  3. readability 库升级为 0.8.1。

## 1.26.6
  1. 添加漫画 扑飞, 古风 的支持。
  2. 添加更多推送书籍的语种。

## 1.26.5
  1. bugfix: 修正1.26.3引入的BUG，此BUG导致排版异常和在电脑上目录跳转到不正确的位置。

## 1.26.4
  1. bugfix: 漫画书籍推送失败，1.26.2引入的bug。

## 1.26.3
  1. 不在用户上传的CSS样式表里面添加命名空间。

## 1.26.2
  1. 支持用户上传CSS样式表文件来定制推送到书籍样式。
  2. 清理失效的书籍。

## 1.26.1
  1. 共享库里新添加的源“置顶”一天。
  2. bugfix: 修正不同账号无法订阅相同URL的BUG。
  
## 1.26
  1. 添加 共享库 特性，所有KindleEar部署的应用都可以使用，并且支持将自己的订阅分享给其他网友。
  2. [网友贡献] 完善CSS，添加响应式设计，适配手机浏览器。
  3. [网友贡献] 添加Bookmarklet。
  4. [网友贡献] 添加多一些漫画支持。
  5. [网友贡献] 添加Sendgrid的邮件发送选项，可以支持发送更多的邮件。

## 1.25.5
  1. skiinder<https://github.com/skiinder>实现在线连载小说的推送。
  2. 导入OPML订阅列表时可选默认按全文还是摘要订阅源导入。

## 1.25.4
  1. 改进手动推送功能，如果仅订阅了自定义RSS，则可以选择推送哪些订阅源。

## 1.25.3
  1. 手动推送支持选择部分书籍(此功能移动到“高级设置|现在投递”)。
  2. 新增移除文本链接选项。
  
## 1.25.2
  1. 更新内置书籍“Economist”的RSS订阅源链接。
  2. bugfix: 修正邮件中转模块推送失败问题。
  3. bugfix: 修正保存到Instapaper功能失效问题。
  
## 1.25.1
  1. 新特性：创建其他账号时可选账号过期期限。
  2. bugfix: 修正漫画模式不压缩图片的问题。

## 1.25
  1. 添加漫画模式和多本漫画书。
  2. 添加对Kindle Voyage和Kindle PaperWhite3的支持。

## 1.24
  1. 新特性：用户可以在GUI界面上直接上传修改封面图片，而且每个账号可单独设置不同的封面图片。
  2. bugfix：修正 反“反盗链”特性 无法下载包含unicode字符的URL的图片的一个bug。
  
## 1.23.5
  1. 新特性：可以在线抓取封面，在自定义py书籍文件中将coverfile属性设置为你自己的抓取函数即可。
  2. bugfix: 修正部分源没有title属性导致推送失败的问题。

## 1.23.3
  1. 增加可在文章后附加网址二维码的功能，方便用手机打开感兴趣的文章继续阅读。 
  2. 增加一个调试功能：发往 debug@xxx.appspotmail.com 的邮件将直接抓取邮件中链接的网页并转发到对应的管理员邮箱。

## 1.23.1
  1. 增强错误处理，规避部分网友部署后无法建立索引的问题。
  2. Bugfix: 修正书籍设置中extra_css无效的问题。
  3. Bugfix: 修正部分全文RSS XML中HTML链接格式不规范导致无法解析出内容的问题。

## 1.23
  1. 添加自动将超长图切割为多个小图片的功能，方便在kindle上看长图。
  2. 进一步优化自动网页解码功能，减小乱码情况。
  3. 修正导入含中文字符URL的OPML文件导致程序崩溃的问题。

## 1.22.3
  1. 修正遇到xml文件中的非法tag（中文tag）导致推送失败的问题。

## 1.22.2
  1. 邮件中转模块升级，在邮件标题中添加 !links则强制抓取文章链接，!article则强制发送文本。
  2. Kindle邮件地址支持多个收件人地址，用分号分隔。

## 1.22.1
  1. 升级内部一些模块。

## 1.22
  1. 用AJAX技术重新实现“我的订阅”前后台，实现不重新加载页面增删订阅或RSS。

## 1.21.1
  1. 增加保存到Instapaper的归档功能。

## 1.21
  1. 增加保存到Pocket的归档功能。

## 1.20.28
  1. 《雪球网》屏蔽了gae的IP段的网络请求，此版本通过中转器获取。

## 1.20.27
  1. 邮件中转功能中引入字符串压缩，支持一封邮件内包含更多链接，以便更好的通过发送链接生成电子书。
  
## 1.20.26
  1. 引入html内嵌base64图像技术，减小http请求数量，提升网页加速效率。
  
## 1.20.25
  1. 增强网页解码，容忍部分字符解码错误，减小了网页乱码可能。

## 1.20.24
  1. 修改选择了合并推送但仅有自定义RSS推送时无封面问题。

## 1.20.23
  1. 去掉标题格式中的 “标题 日/月” 和 “标题 月/日” 格式，因为会导致推送失败。

## 1.20.22
  1. 解决URL中有unicode字符时无法导出OPML文件的文件。
  2. 解决微信公众号获取失败后导致其他书籍推送异常的问题。
  
## 1.20.21
  1. 添加ETAG机制，减小需要的网络流量。

## 1.20.20
  1. 修正部分XML文件获取文件编码失败问题。

## 1.20.19
  1. 修正pycrypto模块加载失败问题。

## 1.20.18
  1. 微信公众号突破一次防爬取 [zhu327](https://github.com/zhu327/rss)。
  2. 通过邮件链接抓取的文章不添加封面。
  
## 1.20.17
  1. 在目录中添加各个源的文章篇数。
  2. 增加对书籍异常的处理，避免一本书籍的异常而影响其他书籍的推送。

## 1.20.16
  1. 增加一个选项，可选择使用网页标题还是feed标题做为文章标题。

## 1.20.15
  1. bugfix: weixinbase部分解码失败问题。

## 1.20.14
  1. bugfix: 喷嚏图啩原RSS地址已经失效，更换为另一个地址。

## 1.20.13
  1. bugfix: 修改有部分RSS将一个图片做为一篇文章（没有html包装）导致推送失败的问题。

## 1.20.12
  1. 导出opml时将url转义。

## 1.20.11
  1. "导入订阅列表"功能支持OPML元素Outline嵌套。

## 1.20.10
  1. bugfix: 修正有部分文章无法提取正文导致推送失败的问题。

## 1.20.9
  1. 增加从opml文件导入订阅列表的功能。
  2. 网友seff增加特性：可选部分书籍单独推送。

## 1.20.8
  1. 网友mcfloundinho增加《共识网》

## 1.20.7
  1. bugfix:修正trigger@appid.appspotmail.com触发投递失败的问题。

## 1.20.6
  1. bugfix:继续修改urlopener处理cookie的一个bug.

## 1.20.5
  1. 增加网友mcfloundinho提供的《南方周末》。
  2. bugfix:修改urlopener处理cookie的一个bug.

## 1.20.4
  1. 更新土耳其语翻译。

## 1.20.3
  1. 修改内置书籍TED渤海湾以适应其网站改版。

## 1.20.2
  1. 针对使用图片延迟加载技术的网页特殊处理，可以获取部分此类网页的图片。

## 1.20.1
  1. 新特性，在合并推送时将各书籍的封面拼贴起来。默认已经开启，如果你使用以前的config.py，请设置DEFAULT_COVER_BV=None，如果不喜欢此效果，可以设置DEFAULT_COVER_BV='cv_bound.jpg'
  2. bugfix: 修正保存到evernote不成功的问题（1.13引入）
  
## 1.20
  1. 增加一个简单的正文提取模块，在readability失败后启用。
  2. 增强的网页解码器，综合考虑http响应头/html文件头/chardet检测结果，效率更高，乱码更少。
  3. 支持需要登陆才能查看文章的网站，请参照FAQ如何使用。
  4. 针对一天推送多次的需求，书籍属性‘oldest_article’大于365则使用*秒*为单位。
  5. 增强的密码安全，加salt然后md5，无法通过密码词典破解，在可接受的代价范围内无法暴力破解。
    （仅新账号启用，如果需要可以删掉admin然后重新登陆就会新建admin账号）
  6. 整理文件夹结构，将相关库都放到lib目录下。
  7. 其他一些小的可用性增强。
  > 升级注意:书籍的fetcharticle()增加了一个参数，如果你定制的书籍使用到此接口，需要修改。
  

config.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-
# KindleEar configuration.
# All configuration variables in this file can be set via environment variables too.
# Values from this file are only used if the environment variables are not present.
#
# 这个文件里面的所有配置变量都可以通过环境变量来设置。
# 程序会优先使用环境变量，只有环境变量不存在时才使用此文件中的数值。
#

APP_ID = "kindleear"
APP_DOMAIN = "https://kindleear.appspot.com"

#Need for google taskqueue only, Refers to <https://cloud.google.com/appengine/docs/locations>
#Find it at Upper right corner of <https://console.cloud.google.com/appengine?project=your_app_id>
#Or get by cmd: gcloud beta app describe
#Two exceptions: europe-west should be europe-west1, us-central should be us-central1
SERVER_LOCATION = "us-central1"

#Choose the database
#Supports: "datastore", "sqlite", "mysql", "postgresql", "cockroachdb", "mongodb", "redis", "pickle"
#DATABASE_URL = "mongodb://127.0.0.1:27017/"
#DATABASE_URL = 'sqlite:////home/ubuntu/site/kindleear/database.db'
#DATABASE_URL = 'sqlite:///database.db'
DATABASE_URL = "sqlite:////data/kindleear.db"

#Select the type of task queue, "gae", "apscheduler", "celery", "rq", ""
TASK_QUEUE_SERVICE = "apscheduler"

#If task queue service is apscheduler, celery, rq
#Options: 'redis://', 'mongodb://', 'sqlite://', 'mysql://', 'postgresql://'
#For apscheduler, it can be 'memory'. (Only if gunicorn have one worker)
#For rq, only 'redis://' is supported
#TASK_QUEUE_BROKER_URL = "redis://127.0.0.1:6379/"
TASK_QUEUE_BROKER_URL = "memory"

#If this option is empty, temporary files will be stored in memory
#Setting this option can reduce memory consumption, supports both relative and absolute paths
KE_TEMP_DIR = "/tmp"

#If online reading is required, this directory is used to store the generated e-books.
EBOOK_SAVE_DIR = ""

#Offline dictionaries for online reading
DICTIONARY_DIR = ""

#If the depolyment plataform supports multi-threads, set this option will boost the download speed
DOWNLOAD_THREAD_NUM = "3"

#If the website allow visitors to signup or not, "yes"|"no"
ALLOW_SIGNUP = "no"

#The secret key for browser session.
SECRET_KEY = "n7ro8QJI1qfe"

#The secret key for starting delivery
DELIVERY_KEY = "cY9gKC"

#The administrator's login name
ADMIN_NAME = "admin"

#You can use this public key or apply for your own key
POCKET_CONSUMER_KEY = "50188-e221424f1c9ed0c010058aef"

#Hide the option 'local (debug)' of 'Send Mail Service' setting or not, "yes"|"no"
HIDE_MAIL_TO_LOCAL = "yes"

#'debug', 'info', 'warning', 'error', 'critical'
LOG_LEVEL = "warning"
